### PR TITLE
Add filtering to transactions

### PR DESCRIPTION
**Pull request by hihuz**
Addresses: https://github.com/kontist/figure-backend/issues/6151

Corresponding backend PR: https://github.com/kontist/figure-backend/pull/6409

So I went with extending the fetch method to allow filter, and also added a "search" helper which does some formatting on a passed string to cover the most generic use case (the one we will need to implement in both webapp and native app).

I think it's nice to have this additional helper for unifying logic in both our apps, and also provide some easy to use method if users don't want to dig too deep into filters

### CODE REVIEW

File: lib/graphql/transaction.ts [hunk 1]
```diff
@@ -1,13 +1,18 @@
...
+  BaseOperator,
   MutationCategorizeTransactionArgs,
   Query,
   Transaction as TransactionModel,
+  TransactionFilter,
   TransactionsConnectionEdge,
 } from "./schema";
 import { FetchOptions, Subscription, SubscriptionType } from "./types";
 
+const MAX_SEARCH_QUERY_LENGTH = 200;
```
Comment (hihuz) [inline lib/graphql/transaction.ts]: What do you think of truncating user query if too long? is 200 chars proper?

Do you also think I should perform some other sanitization operations before sending the stuff to the backend? 🤔

File: lib/graphql/transaction.ts [hunk 2]
```diff
@@ -120,4 +140,28 @@ export class Transaction extends IterableModel<TransactionModel> {
...
+    const searchTerms = searchQuery.slice(0, MAX_SEARCH_QUERY_LENGTH).split(" ");
+
+    const filter: TransactionFilter = {
+      iban_likeAny: searchTerms,
+      name_likeAny: searchTerms,
+      operator: BaseOperator.Or,
+      purpose_likeAny: searchTerms,
+    };
+
+    const amountRegex = /^-?\d+([,.]\d{1,2})?$/;
```
Comment (hihuz) [inline lib/graphql/transaction.ts]: So, I want to match: `1234`, `-1234`, `1.23`, `12.1`, `1,23`, `12,34` etc. etc.

I don't want to match shit like `.10`, `10.`, `10,` but maybe I should? 🤔

File: lib/graphql/types.ts [hunk 3]
```diff
@@ -1,22 +1,12 @@
...
-  last?: number;
-  before?: string | null;
-  after?: string | null;
-};
-
-export type TransferFetchOptions = {
-  type: TransferType;
-  where?: TransfersConnectionFilter;
-} & FetchOptions;
+export type FetchOptions = AccountTransactionsArgs | AccountTransfersArgs;
```
Comment (hihuz) [inline lib/graphql/types.ts]: what do you think of this? 🤔 it's not strictly needed but we were once again creating some custom types that we can just get from the Schema

File: lib/graphql/transaction.ts (hunk 2 above)
Comment (p-janik) [reply lib/graphql/transaction.ts]: I would vote for obvious ones only, so no .10 💩

File: lib/graphql/transaction.ts (hunk 1 above)
Comment (p-janik) [reply lib/graphql/transaction.ts]: 200 I a lot and should be enough, I don't think that user would type that much or paste that long strings

File: lib/graphql/transaction.ts [hunk 4]
```diff
@@ -91,6 +98,19 @@ export class Transaction extends IterableModel<TransactionModel> {
     return new ResultPage(this, transactions, pageInfo, args);
   }
 
+  /**
+   * Fetches first 50 matching transactions for a given user input
```
Comment (p-janik) [inline lib/graphql/transaction.ts]: `fetches up to first 50`?

File: lib/graphql/transaction.ts [hunk 5]
```diff
@@ -120,4 +140,28 @@ export class Transaction extends IterableModel<TransactionModel> {
...
+      name_likeAny: searchTerms,
+      operator: BaseOperator.Or,
+      purpose_likeAny: searchTerms,
+    };
+
+    const amountRegex = /^-?\d+([,.]\d{1,2})?$/;
+    const amountTerms = searchTerms
+      .filter((term) => amountRegex.test(term))
+      .reduce((terms: number[], term: string): number[] => {
+        const amountInCents = parseFloat(term.replace(",", ".")) * 100;
```
Comment (p-janik) [inline lib/graphql/transaction.ts]: what will happen if i provide `33.91`? :P

Comment (hihuz) [reply lib/graphql/transaction.ts]: oh shit nice catch 🤦‍♂ what if I Math.round afterwards? it should work for all cases, I'm trying to come up with a counter example

File: lib/graphql/types.ts (hunk 3 above)
Comment (johannespfeiffer) [reply lib/graphql/types.ts]: Not a fan of this change (`FetchOptions`):
1. We will need to adjust this for every new collection.
2. It allows using `transaction.fetchAll` with `args` of type `AccountTransfersArgs`.

I do like that we can get rid of `TransferFetchOptions`.

File: lib/graphql/transaction.ts (hunk 1 above)
Comment (johannespfeiffer) [reply lib/graphql/transaction.ts]: 200 characters should be enough.

File: lib/graphql/transaction.ts [hunk 6]
```diff
@@ -91,6 +98,19 @@ export class Transaction extends IterableModel<TransactionModel> {
     return new ResultPage(this, transactions, pageInfo, args);
   }
 
+  /**
+   * Fetches first 50 matching transactions for a given user input
+   * It will consider case insentsitive like matches for amount,
```
Comment (johannespfeiffer) [inline lib/graphql/transaction.ts]: nit: typo "insentsitive"

File: tests/graphql/transaction.spec.ts [hunk 7]
```diff
@@ -75,6 +81,29 @@ describe("Transaction", () => {
...
+          operator: BaseOperator.And,
+          amount_gt: 1000,
+          name_like:"SaNtA"
+        }
+      });
+
+      expect(results.items).to.deep.equal([
+        firstTransaction,
+        thirdTransaction,
+      ]);
```
Comment (johannespfeiffer) [inline tests/graphql/transaction.spec.ts]: Can we extend the expectation a bit? E.g.
```
      sinon.assert.calledWithExactly(stub, sinon.match.string, {
        filter: {
          operator: BaseOperator.And,
          amount_gt: 1001,
          name_like: "SaNtA"
        }
      });
```

Without it the test runs fine even without the filter.

File: lib/graphql/transaction.ts (hunk 5 above)
Comment (hihuz) [reply lib/graphql/transaction.ts]: I think `Math.round` should work in all cases: 84d001fa0a21b1e74b87c5e99108e8a05d855c7a

File: lib/graphql/types.ts (hunk 3 above)
Comment (hihuz) [reply lib/graphql/types.ts]: Makes total sense, I reverted that part of my changes: 0032192b41d45d1f6fd152fc7e70fa7bc1768709

File: lib/graphql/transaction.ts (hunk 6 above)
Comment (hihuz) [reply lib/graphql/transaction.ts]: Thank you 🙂 c8d7e7dec33255d94d7e49a1b6c3d54ac3a484f9

File: lib/graphql/transaction.ts (hunk 4 above)
Comment (hihuz) [reply lib/graphql/transaction.ts]: Good idea, c8d7e7dec33255d94d7e49a1b6c3d54ac3a484f9

File: tests/graphql/transaction.spec.ts (hunk 7 above)
Comment (hihuz) [reply tests/graphql/transaction.spec.ts]: Um good point 😂 test was pretty useless. What about this assertion 9aec03fb1bc6c0f27a521c4f75815ea354e38e1a ?

Comment (johannespfeiffer) [reply tests/graphql/transaction.spec.ts]: 👍 thank you

**Pull request merged**